### ISSUE TITLE
Serialization model for tB

### ISSUE THREAD

**Issue by bclothier** [WaynePhillipsEA/twinbasic#453]
This came up as a discussion in this GUI upgrade progress:
>The deserialization needs some more thought to allow for customization.  At the moment, it's a black box, with the Deserialize call handling all the nitty gritty details, but not offering any customization.  I'm certainly open to ideas on this.

_Originally posted by @WaynePhillipsEA in https://github.com/WaynePhillipsEA/twinbasic/discussions/451#discussioncomment-1384069_

I am aware in VB6 there's the property bag but if I'm not mistaken, that doesn't give the developers full control over the serialization of an object. In VBA there's absolutely nothing to make it easy to serialize/deserialize an object. For those reasons, several items may be inaccessible (e.g. saved in binary format in a FRX file, not visible in code, etc.) which all are bad for source code control. Since we want tB to be friendly toward source code control and encourage open documentation of the object's internals, it might be good to discuss how tB should support serialization of the objects and do so that is easily accessible to the developers and can be extended for edge cases. 

In .NET, they solved the problem by providing [`Serializable` attribute](https://docs.microsoft.com/en-us/dotnet/api/system.serializableattribute?view=net-5.0) and [`ISerializable` interface](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.iserializable?view=net-5.0). It is my understanding that:
* All objects that have a design-time elements must implement this interface or equivalent (usually implicitly)
* .NET has several different serialization system (e.g. [`DataContract`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.datacontractattribute?view=net-5.0). This is possible because they have different implementation of reflecting over a type in order to serialize/deserialize.
* .NET provides default serializers so that we don't have to write our own serializer by hand. In tB's case, it probably makes sense to provide a JSON serializer given that it's working with JSON.

Because VBx does not have strong reflection capability, writing code for serialization is very tedious and would require several lines to manually construct/deconstruct the values to the stream and vice versa. In .NET, reflection API helps makes it more generic and thus easy to implement & extend. I think that strongly implies that for us to have the same in tB, tB needs to have a good reflection API (which can be just be a VB-friendly facade over the `ITypeInfo` that is already used internally). One consideration is the need to be able to access the private members as some may not be able to fully serialize/deserialize if the private members are skipped. For that reason, .NET reflection allows one to access the private members to cover those objects that would otherwise be non-serializable. Making those members public is not always a desirable option. 

To help minimize the reinvention of wheel, I think tB would be well to copy some (not necessarily all) the prior art as exhibited by .NET in suporting serialization of objects.

**Comment by WaynePhillipsEA** [comment]
The main thing I want to mention is that the built-in serialization/deserialization is very efficient.  For example, the runtime deserializer is binary based, not JSON based, and doesn't use property bags or any form of named-referencing for associating values with properties (it references field by offsets determined at compile-time, and vtable indexes for property setters). 

I'm all for allowing customized serialization/deserialization as suggested here, but the built-in automatic serialization that is coming with the CustomControls should be considered a good default implementation, and only use custom serialization if really needed.

For custom serialization, I'm actually wondering if it might just be easier and more familiar to use the PropertyBag system used in traditional ActiveX control designs on VB6, rather the .NET way.  The property bags could then just persisted as extensions on the built-in automatic serialization info, allowing to use both the built-in efficient serialization, with any 'extra' serialization requirements provided through property bags.

**Comment by bclothier** [comment]
I agree that the property bag system is a simple way to customize the serialization behavior, but if I am not mistaken, you still don't get to save or read from a text file and create a class from it. This is only done by VB6 implicitly as part of the loading and you don't even get access to the file. 

In this case, we're talking about arbitrarily serializing a class to a specified file in response to our code asking for it. I might want to store a state so that it can be later reverted. Under VBx, that would mean I'd need to write a text file, write reams of code to read the text file, assign to the properties to construct the class, write more reams of code to write the text file from the class. With the .NET serialization model, it's just a matter of decorating a class with an attribute and woosh! I now can serialize the class as a XML, JSON, binary, or even a custom file format using my custom serializer.

That is a different use case from the one that custom controls require since custom controls require it for design time experience, but the above would be a part of the runtime experience, not design time. 

I hope that clarifies.

**Comment by wqweto** [comment]
JFYI, in VB6 public classes on ActiveX DLL/EXE can be marked Persistable so these get a pair of extra base class events (inherit different base class)
```
Private Sub Class_ReadProperties(PropBag As PropertyBag)

End Sub

Private Sub Class_WriteProperties(PropBag As PropertyBag)

End Sub
```
"Calling" these events happens like this
```
Private Sub Form_Load()
    Dim o As MyClass
    Set o = New MyClass
    
    With New PropertyBag
        .WriteProperty "MyObj", o, Nothing
    End With
    '...
    With New PropertyBag
        .Contents = PersistedData
        Set o = .ReadProperty("MyObj", Nothing)
    End With
End Sub
```
So property bag serializatrion **is** recursive in a sense.
